src/twisted/trial/reporter.py

```diff
@@ -497,35 +517,105 @@ def cleanupErrors(self, errs):
             BrokenTestCaseWarning,
         )
 
+    def _trimRunnerFrames(self, frames):
+        def frameInfo(f):
+            return f[0], os.path.splitext(os.path.basename(f[1]))[0]
+
+        syncCase = [("_run", "_synctest")]
+
+        if len(frames) < 1:
+            return frames
+
+        frame0 = frameInfo(frames[0])
+        if [frame0] == syncCase:
+            return frames[1:]
+
+        if len(frames) < 2:
+            return frames
+
+        asyncCase = [
+            ("_inlineCallbacks", "defer"),
+            ("_runCorofnWithWarningsSuppressed", "_asynctest"),
+        ]
+
+        frame1 = frameInfo(frames[1])
+        if [frame0, frame1] == asyncCase:
+            return frames[2:]
+
+        if len(frames) < 3:
+            return frames
+
+        asyncShimCvarCase = [
+            ("_inlineCallbacks", "defer"),
+            _cvarFrame,
+            ("_runCorofnWithWarningsSuppressed", "_asynctest"),
+        ]
+
+        frame2 = frameInfo(frames[2])
+        if [frame0, frame1, frame2] == asyncShimCvarCase:
+            return frames[3:]
+
+        if len(frames) < 4:
+            return frames
+
+        asyncFailureCase = [
+            ("_inlineCallbacks", "defer"),
+            ("throwExceptionIntoGenerator", "failure"),
+            ("_runCorofnWithWarningsSuppressed", "_asynctest"),
+            ("_runCallbacks", "defer"),
+        ]
+
+        frame2 = frameInfo(frames[2])
+        frame3 = frameInfo(frames[3])
+        if [frame0, frame1, frame2, frame3] == asyncFailureCase:
+            return frames[4:]
+
+        if len(frames) < 5:
+            return frames
+
+        asyncFailureShimCvarCase = [
+            ("_inlineCallbacks", "defer"),
+            _cvarFrame,
+            ("throwExceptionIntoGenerator", "failure"),
+            ("_runCorofnWithWarningsSuppressed", "_asynctest"),
+            ("_runCallbacks", "defer"),
+        ]
+
+        frame4 = frameInfo(frames[4])
+        if [frame0, frame1, frame2, frame3, frame4] == asyncFailureShimCvarCase:
+            return frames[5:]
+
+        return frames
+
     def _trimFrames(self, frames):
```

Comment:
Gosh this function is getting enormously complicated, and it didn't start simple. I kind of want it to live in its own module and:

1. abstract this stuff a bit so you can arbitrarily say what "irrelevant" frames you want to throw away
2. shows you a `... X frames elided ...` message by default so you know *something* is going on
3. make a cmdline option for debugging weird cases that *doesn't* do any of this traceback stripping

Reply:
Pytest and weukzeug hide frames containing a `__tracebackhide__` or `__traceback_hide__` local

Reply:
how about the `@_extraneous` decorator?

Reply:
@glyph I'm inclined to agree, eg. I'd like those 3 features also. However this frame sniffing is at my limit of understanding of Twisted internals and so I don't know how to proceed here.

Reply:
For now let's just stick this into a separate file so that it's easier to come back to as an isolated bit of functionality in a follow-up.